cunumeric/deferred.py

```diff
@@ -305,37 +305,278 @@ def get_scalar_array(self):
+        return store_copy, store_copy.base
+
+    def _create_indexing_array(self, key, is_set=False):
+        store = self.base
+        rhs = self
+        # the index where the first index_array is passed to the [] operator
+        start_index = -1
         if isinstance(key, tuple):
+            key = self._unpack_ellipsis(key, self.ndim)
+            shift = 0
+            last_index = self.ndim
+            # in case when index arrays are passed in the scaterred way,
+            # we need to transpose original array so all index arrays
+            # are close to each other
+            transpose_needed = False
+            transpose_indices = tuple()
+            # since we can't call Copy operation on transformed Store, after
+            # the transformation, we need to return a copy
             tuple_of_arrays = ()
-            for k in key:
-                if not isinstance(k, NumPyThunk):
-                    raise NotImplementedError(
-                        "need support for mixed advanced indexing"
+            index_map = []
+
+            # First, we need to check if transpose is needed
+            for dim, k in enumerate(key):
+                if np.isscalar(k) or isinstance(k, NumPyThunk):
+                    if start_index == -1:
+                        start_index = dim
+                    transpose_indices += (dim,)
+                    transpose_needed = transpose_needed or (
+                        (dim - last_index) > 1
+                    )
+                    last_index = dim
+
+            if transpose_needed:
+                start_index = 0
+                post_indices = tuple(
+                    i for i in range(store.ndim) if i not in transpose_indices
+                )
+                transpose_indices += post_indices
+                store = store.transpose(transpose_indices)
+                index_map = list(transpose_indices)
+                count = 0
+                for i in transpose_indices:
+                    index_map[i] = count
+                    count += 1
+            else:
+                index_map = tuple(range(len(key)))
+
+            for d, k in enumerate(key):
+                dim = index_map[d]
+                if np.isscalar(k):
+                    if k < 0:
+                        k += store.shape[dim + shift]
+                    store = store.project(dim + shift, k)
+                    shift -= 1
+                elif k is np.newaxis:
+                    store = store.promote(dim + shift, 1)
+                elif isinstance(k, slice):
+                    store = store.slice(dim + shift, k)
+                elif isinstance(k, NumPyThunk):
+                    if k.dtype == np.bool:
+                        if k.shape[0] != store.shape[dim]:
+                            raise ValueError(
+                                "boolean index did not match "
+                                "indexed array along dimension  "
```

Comment:
I think it's going to be hard to provide a good error message here, because we may have shuffled the arrays. I think we can just say "shape of boolean index did not match indexed array", without specifying the dimension.

Reply:
sure